File: app/src/main/java/com/arjanvlek/oxygenupdater/notifications/DelayedPushNotificationDisplayer.java
package com.arjanvlek.oxygenupdater.notifications;

import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.job.JobParameters;
import android.app.job.JobService;
import android.content.Intent;
import android.os.Build;
import android.support.v4.app.NotificationCompat;

import com.arjanvlek.oxygenupdater.ApplicationData;
import com.arjanvlek.oxygenupdater.R;
import com.arjanvlek.oxygenupdater.internal.OxygenUpdaterException;
import com.arjanvlek.oxygenupdater.internal.Utils;
import com.arjanvlek.oxygenupdater.internal.logger.Logger;
import com.arjanvlek.oxygenupdater.news.NewsActivity;
import com.arjanvlek.oxygenupdater.settings.SettingsManager;
import com.arjanvlek.oxygenupdater.views.MainActivity;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.IOException;
import java.util.Locale;
import java.util.Map;

import static android.app.Notification.DEFAULT_ALL;
import static android.app.Notification.PRIORITY_HIGH;
import static com.arjanvlek.oxygenupdater.ApplicationData.LOCALE_DUTCH;
import static com.arjanvlek.oxygenupdater.news.NewsActivity.INTENT_NEWS_ITEM_ID;
import static com.arjanvlek.oxygenupdater.news.NewsActivity.INTENT_START_WITH_AD;
import static com.arjanvlek.oxygenupdater.notifications.NotificationElement.DEVICE_NAME;
import static com.arjanvlek.oxygenupdater.notifications.NotificationElement.DUTCH_MESSAGE;
import static com.arjanvlek.oxygenupdater.notifications.NotificationElement.ENGLISH_MESSAGE;
import static com.arjanvlek.oxygenupdater.notifications.NotificationElement.NEWS_ITEM_ID;
import static com.arjanvlek.oxygenupdater.notifications.NotificationElement.NEW_DEVICE_NAME;
import static com.arjanvlek.oxygenupdater.notifications.NotificationElement.NEW_VERSION_NUMBER;
import static com.arjanvlek.oxygenupdater.notifications.NotificationElement.TYPE;
import static com.arjanvlek.oxygenupdater.settings.SettingsManager.PROPERTY_RECEIVE_GENERAL_NOTIFICATIONS;
import static com.arjanvlek.oxygenupdater.settings.SettingsManager.PROPERTY_RECEIVE_NEWS_NOTIFICATIONS;
import static com.arjanvlek.oxygenupdater.settings.SettingsManager.PROPERTY_RECEIVE_NEW_DEVICE_NOTIFICATIONS;
import static com.arjanvlek.oxygenupdater.settings.SettingsManager.PROPERTY_RECEIVE_SYSTEM_UPDATE_NOTIFICATIONS;

/**
 * Oxygen Updater - © 2018 Arjan Vlek
 */
public class DelayedPushNotificationDisplayer extends JobService {

    public static final int NEW_DEVICE_NOTIFICATION_ID = 10010;
    public static final int NEW_UPDATE_NOTIFICATION_ID = 20020;
    public static final int GENERIC_NOTIFICATION_ID = 30030;
    public static final int NEWS_NOTIFICATION_ID = 50050;
    public static final int UNKNOWN_NOTIFICATION_ID = 40040;

    public static final String KEY_NOTIFICATION_CONTENTS = "notification-contents";
    private static final String TAG = "DelayedPushNotificationDisplayer";

    @Override
    public boolean onStartJob(JobParameters params) {
        if (params == null || !params.getExtras().containsKey(KEY_NOTIFICATION_CONTENTS) || !(getApplication() instanceof ApplicationData)) {
            return exit(params, false);
        }

        SettingsManager settingsManager = new SettingsManager(getApplication());

        // Get notification contents of FCM.
        TypeReference<Map<String, String>> notificationContentsTypeRef = new TypeReference<Map<String, String>>() {
        };
        Map<String, String> messageContents;

        String notificationContentsJson = params.getExtras().getString(KEY_NOTIFICATION_CONTENTS);
        try {
            messageContents = new ObjectMapper().readValue(notificationContentsJson, notificationContentsTypeRef);
        } catch (IOException e) {
            Logger.logError(TAG, new OxygenUpdaterException("Failed to read notification contents from JSON string (" + notificationContentsJson + ")"));
            return exit(params, false);
        }
        NotificationType notificationType = NotificationType.valueOf(messageContents.get(TYPE.toString()));
        NotificationCompat.Builder builder = null;

        switch (notificationType) {
            case NEW_DEVICE:
                if (!settingsManager.getPreference(PROPERTY_RECEIVE_NEW_DEVICE_NOTIFICATIONS, true)) {
                    return exit(params, true);
                }
                builder = getBuilderForNewDeviceNotification(messageContents.get(NEW_DEVICE_NAME.toString()));
                break;
            case NEW_VERSION:
                if (!settingsManager.getPreference(PROPERTY_RECEIVE_SYSTEM_UPDATE_NOTIFICATIONS, true)) {
                    return exit(params, true);
                }
                builder = getBuilderForNewVersionNotification(messageContents.get(DEVICE_NAME.toString()), messageContents.get(NEW_VERSION_NUMBER.toString()));
                break;
            case GENERAL_NOTIFICATION:
                if (!settingsManager.getPreference(PROPERTY_RECEIVE_GENERAL_NOTIFICATIONS, true)) {
                    return exit(params, true);
                }

                String message;
                switch (Locale.getDefault().getDisplayLanguage()) {
                    case LOCALE_DUTCH:
                        message = messageContents.get(DUTCH_MESSAGE.toString());
                        break;
                    default:
                        message = messageContents.get(ENGLISH_MESSAGE.toString());
                        break;
                }
                builder = getBuilderForGeneralServerNotificationOrNewsNotification(message);
                break;
            case NEWS:
                if (!settingsManager.getPreference(PROPERTY_RECEIVE_NEWS_NOTIFICATIONS, true)) {
                    return exit(params, true);
                }

                String newsMessage;
                switch (Locale.getDefault().getDisplayLanguage()) {
                    case LOCALE_DUTCH:
                        newsMessage = messageContents.get(DUTCH_MESSAGE.toString());
                        break;
                    default:
                        newsMessage = messageContents.get(ENGLISH_MESSAGE.toString());
                        break;
                }
                builder = getBuilderForGeneralServerNotificationOrNewsNotification(newsMessage);
                break;

        }
        if (builder == null) {
            Logger.logError(TAG, new OxygenUpdaterException("Failed to instantiate notificationBuilder. Can not display push notification!"));
            return exit(params, false);
        }

        builder.setContentIntent(getNotificationIntent(notificationType, messageContents));
        builder.setVisibility(Notification.VISIBILITY_PUBLIC);
        builder.setPriority(PRIORITY_HIGH);
        builder.setDefaults(DEFAULT_ALL);
        builder.setAutoCancel(true);

        int notificationId = getNotificationId(notificationType);
        Notification notification = builder.build();

        NotificationManager notificationManager = (NotificationManager) Utils.getSystemService(this, NOTIFICATION_SERVICE);

        if (notificationManager == null) {
            Logger.logError(TAG, new OxygenUpdaterException("Notification Manager service is not available"));
            return exit(params, false);
        }

        notificationManager.notify(notificationId, notification);
        return exit(params, true);
    }

    private int getNotificationId(NotificationType type) {
        switch (type) {
            case NEW_DEVICE:
                return NEW_DEVICE_NOTIFICATION_ID;
            case NEW_VERSION:
                return NEW_UPDATE_NOTIFICATION_ID;
            case GENERAL_NOTIFICATION:
                return GENERIC_NOTIFICATION_ID;
            case NEWS:
                return NEWS_NOTIFICATION_ID;
            default:
                return UNKNOWN_NOTIFICATION_ID;
        }
    }

    @Override
    public boolean onStopJob(JobParameters params) {
        return true;
    }

    private NotificationCompat.Builder getBuilderForGeneralServerNotificationOrNewsNotification(String message) {
        return getNotificationBuilder()
                .setSmallIcon(R.drawable.ic_stat_notification_general)
                .setStyle(new NotificationCompat.BigTextStyle()
                        .bigText(message))
                .setContentTitle(getString(R.string.app_name))
                .setContentText(message);
    }

    private NotificationCompat.Builder getBuilderForNewDeviceNotification(String newDeviceName) {
        String message = getString(R.string.notification_new_device, newDeviceName);

        return getNotificationBuilder()
                .setSmallIcon(R.drawable.ic_stat_notification_new_device)
                .setStyle(new NotificationCompat.BigTextStyle()
                        .bigText(message)
                        .setSummaryText(getString(R.string.notification_new_device_short)))
                .setContentTitle(getString(R.string.app_name))
                .setContentText(message);
    }

    private NotificationCompat.Builder getBuilderForNewVersionNotification(String deviceName, String versionNumber) {
        String message = getString(R.string.notification_version, versionNumber, deviceName);
        return getNotificationBuilder()
                .setSmallIcon(R.drawable.ic_stat_notification_new_version)
                .setStyle(new NotificationCompat.BigTextStyle()
                        .bigText(message))
                .setWhen(System.currentTimeMillis())
                .setContentTitle(getString(R.string.notification_version_title))
                .setContentText(message);
    }


    private NotificationCompat.Builder getNotificationBuilder() {
        if (Build.VERSION.SDK_INT >= 26) {
            return new NotificationCompat.Builder(this, ApplicationData.PUSH_NOTIFICATION_CHANNEL_ID);
        } else {
            //noinspection deprecation - Only runs on older Android versions.
            return new NotificationCompat.Builder(this);
        }
    }

    private PendingIntent getNotificationIntent(NotificationType notificationType, Map<String, String> messageContents) {
        PendingIntent contentIntent;

        if (notificationType == NotificationType.NEWS) {
            Intent newsIntent = new Intent(this, NewsActivity.class);
            newsIntent.putExtra(INTENT_NEWS_ITEM_ID, Long.valueOf(messageContents.get(NEWS_ITEM_ID.toString())));
            newsIntent.putExtra(INTENT_START_WITH_AD, true);
            contentIntent = PendingIntent.getActivity(this, 0, newsIntent, 0);
        } else {
            contentIntent = PendingIntent.getActivity(this, 0, new Intent(this, MainActivity.class), 0);
        }

        return contentIntent;
    }

    private boolean exit(JobParameters parameters, boolean success) {
        jobFinished(parameters, !success);
        return success;
    }
}

File: app/src/main/java/com/arjanvlek/oxygenupdater/notifications/DelayedPushNotificationDisplayer.kt
package com.arjanvlek.oxygenupdater.notifications

import android.app.Notification
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.job.JobParameters
import android.app.job.JobService
import android.content.Context
import android.content.Intent
import android.os.Build
import androidx.core.app.NotificationCompat
import com.arjanvlek.oxygenupdater.ApplicationData
import com.arjanvlek.oxygenupdater.R
import com.arjanvlek.oxygenupdater.internal.OxygenUpdaterException
import com.arjanvlek.oxygenupdater.internal.Utils
import com.arjanvlek.oxygenupdater.internal.i18n.Locale
import com.arjanvlek.oxygenupdater.internal.i18n.Locale.NL
import com.arjanvlek.oxygenupdater.internal.logger.Logger.logError
import com.arjanvlek.oxygenupdater.news.NewsActivity
import com.arjanvlek.oxygenupdater.notifications.NotificationElement.DEVICE_NAME
import com.arjanvlek.oxygenupdater.notifications.NotificationElement.DUTCH_MESSAGE
import com.arjanvlek.oxygenupdater.notifications.NotificationElement.ENGLISH_MESSAGE
import com.arjanvlek.oxygenupdater.notifications.NotificationElement.NEW_DEVICE_NAME
import com.arjanvlek.oxygenupdater.notifications.NotificationElement.NEW_VERSION_NUMBER
import com.arjanvlek.oxygenupdater.notifications.NotificationElement.TYPE
import com.arjanvlek.oxygenupdater.notifications.NotificationType.GENERAL_NOTIFICATION
import com.arjanvlek.oxygenupdater.notifications.NotificationType.NEWS
import com.arjanvlek.oxygenupdater.notifications.NotificationType.NEW_DEVICE
import com.arjanvlek.oxygenupdater.notifications.NotificationType.NEW_VERSION
import com.arjanvlek.oxygenupdater.settings.SettingsManager
import com.arjanvlek.oxygenupdater.views.MainActivity
import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import java.io.IOException

/**
 * Oxygen Updater - © 2018 Arjan Vlek
 */
class DelayedPushNotificationDisplayer : JobService() {

    override fun onStartJob(params: JobParameters?): Boolean {
        if (params == null || !params.extras.containsKey(KEY_NOTIFICATION_CONTENTS) || application !is ApplicationData) {
            return exit(params, false)
        }

        val settingsManager = SettingsManager(application)

        // Get notification contents of FCM.
        val notificationContentsTypeRef: TypeReference<Map<String, String>> = object : TypeReference<Map<String, String>>() {}
        val messageContents: Map<String, String>
        val notificationContentsJson = params.extras.getString(KEY_NOTIFICATION_CONTENTS)

        messageContents = try {
            jacksonObjectMapper().readValue(notificationContentsJson, notificationContentsTypeRef)
        } catch (e: IOException) {
            logError(TAG, OxygenUpdaterException("Failed to read notification contents from JSON string ($notificationContentsJson)"))
            return exit(params, false)
        }

        val notificationType = NotificationType.valueOf(messageContents[TYPE.toString()] ?: "")

        val builder: NotificationCompat.Builder? = when (notificationType) {
            NEW_DEVICE -> {
                if (!settingsManager.getPreference(SettingsManager.PROPERTY_RECEIVE_NEW_DEVICE_NOTIFICATIONS, true)) {
                    return exit(params, true)
                }

                getBuilderForNewDeviceNotification(messageContents[NEW_DEVICE_NAME.toString()])
            }
            NEW_VERSION -> {
                if (!settingsManager.getPreference(SettingsManager.PROPERTY_RECEIVE_SYSTEM_UPDATE_NOTIFICATIONS, true)) {
                    return exit(params, true)
                }

                getBuilderForNewVersionNotification(messageContents[DEVICE_NAME.toString()], messageContents[NEW_VERSION_NUMBER.toString()])
            }
            GENERAL_NOTIFICATION -> {
                if (!settingsManager.getPreference(SettingsManager.PROPERTY_RECEIVE_GENERAL_NOTIFICATIONS, true)) {
                    return exit(params, true)
                }

                val message = if (Locale.locale == NL) messageContents[DUTCH_MESSAGE.toString()] else messageContents[ENGLISH_MESSAGE.toString()]

                getBuilderForGeneralServerNotificationOrNewsNotification(message)
            }
            NEWS -> {
                if (!settingsManager.getPreference(SettingsManager.PROPERTY_RECEIVE_NEWS_NOTIFICATIONS, true)) {
                    return exit(params, true)
                }

                val newsMessage = if (Locale.locale == NL) messageContents[DUTCH_MESSAGE.toString()] else messageContents[ENGLISH_MESSAGE.toString()]

                getBuilderForGeneralServerNotificationOrNewsNotification(newsMessage)
            }
        }

        if (builder == null) {
            logError(TAG, OxygenUpdaterException("Failed to instantiate notificationBuilder. Can not display push notification!"))
            return exit(params, false)
        }

        builder.setContentIntent(getNotificationIntent(notificationType, messageContents))
            .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
            .setPriority(Notification.PRIORITY_HIGH)
            .setDefaults(Notification.DEFAULT_ALL)
            .setAutoCancel(true)

        val notificationId = getNotificationId(notificationType)
        val notification = builder.build()
        val notificationManager: NotificationManager? = Utils.getSystemService(this, Context.NOTIFICATION_SERVICE) as NotificationManager

        if (notificationManager == null) {
            logError(TAG, OxygenUpdaterException("Notification Manager service is not available"))
            return exit(params, false)
        }

        notificationManager.notify(notificationId, notification)

        return exit(params, true)
    }

    override fun onStopJob(params: JobParameters): Boolean {
        return true
    }

    private fun getNotificationId(type: NotificationType): Int {
        return when (type) {
            NEW_DEVICE -> NEW_DEVICE_NOTIFICATION_ID
            NEW_VERSION -> NEW_UPDATE_NOTIFICATION_ID
            GENERAL_NOTIFICATION -> GENERIC_NOTIFICATION_ID
            NEWS -> NEWS_NOTIFICATION_ID
            else -> UNKNOWN_NOTIFICATION_ID
        }
    }

    private fun getBuilderForGeneralServerNotificationOrNewsNotification(message: String?): NotificationCompat.Builder {
        return notificationBuilder
            .setSmallIcon(R.drawable.oxygen_updater)
            .setStyle(NotificationCompat.BigTextStyle().bigText(message))
            .setContentTitle(getString(R.string.app_name))
            .setContentText(message)
    }

    private fun getBuilderForNewDeviceNotification(newDeviceName: String?): NotificationCompat.Builder {
        val message = getString(R.string.notification_new_device, newDeviceName)

        return notificationBuilder
            .setSmallIcon(R.drawable.new_text)
            .setStyle(NotificationCompat.BigTextStyle().bigText(message).setSummaryText(getString(R.string.notification_new_device_short)))
            .setContentTitle(getString(R.string.app_name))
            .setContentText(message)
    }

    private fun getBuilderForNewVersionNotification(deviceName: String?, versionNumber: String?): NotificationCompat.Builder {
        val message = getString(R.string.notification_version, versionNumber, deviceName)

        return notificationBuilder
            .setSmallIcon(R.drawable.new_text)
            .setStyle(NotificationCompat.BigTextStyle().bigText(message))
            .setWhen(System.currentTimeMillis())
            .setContentTitle(getString(R.string.notification_version_title))
            .setContentText(message)
    }

    private val notificationBuilder: NotificationCompat.Builder
        get() = if (Build.VERSION.SDK_INT >= 26) {
            NotificationCompat.Builder(this, ApplicationData.PUSH_NOTIFICATION_CHANNEL_ID)
        } else {
            NotificationCompat.Builder(this)
        }

    private fun getNotificationIntent(notificationType: NotificationType, messageContents: Map<String, String>): PendingIntent {
        val contentIntent: PendingIntent
        contentIntent = if (notificationType == NEWS) {
            val newsIntent = Intent(this, NewsActivity::class.java)
                .putExtra(NewsActivity.INTENT_NEWS_ITEM_ID, messageContents[NotificationElement.NEWS_ITEM_ID.toString()]?.toLong())
                .putExtra(NewsActivity.INTENT_START_WITH_AD, true)

            PendingIntent.getActivity(this, 0, newsIntent, 0)
        } else {
            PendingIntent.getActivity(this, 0, Intent(this, MainActivity::class.java), 0)
        }
        return contentIntent
    }

    private fun exit(parameters: JobParameters?, success: Boolean): Boolean {
        jobFinished(parameters, !success)
        return success
    }

    companion object {
        const val NEW_DEVICE_NOTIFICATION_ID = 10010
        const val NEW_UPDATE_NOTIFICATION_ID = 20020
        const val GENERIC_NOTIFICATION_ID = 30030
        const val NEWS_NOTIFICATION_ID = 50050
        const val UNKNOWN_NOTIFICATION_ID = 40040
        const val KEY_NOTIFICATION_CONTENTS = "notification-contents"
        private const val TAG = "DelayedPushNotificationDisplayer"
    }
}
